From 1dbe3317a03f37988e12cf799ae3c97a74d472fc Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Sun, 26 Apr 2026 20:59:23 +0000
Subject: [PATCH 160/280] WIP: drm/bridge: dw-hdmi: use limited-range RGB to
 YCbCr CSC coeffs

dw_hdmi_update_csc_coeffs() uses the full-range RGB to YCbCr CSC
matrices even though the AVI infoframe advertises limited-range YCbCr
output. This makes the CSC programming inconsistent with the output
range signalled to the sink.

DW HDMI TX databook mentions:

  The coefficients used in the CSC matrix use only 15 bits for the
  internal computations. Coefficients are represented in 2's
  complementary format and stored in two registers:

  csc_coef_*_lsb[7:0]: coefficient bits 7 to 0
  csc_coef_*_msb[7]: spare bit
  csc_coef_*_msb[6:0]: coefficient bits 14 to 8

Add dedicated BT.601 and BT.709 RGB full-range to YCbCr limited-range
coefficient tables and use them for RGB to YCbCr conversion. Recalculate
the coefficients to fit the DW HDMI 15-bit CSC format.

Also update the RGB full-range to RGB limited-range coefficients to the
rounded 15-bit values expected by the hardware.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 drivers/gpu/drm/bridge/synopsys/dw-hdmi.c | 32 +++++++++++++++++------
 1 file changed, 24 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
index 4777bfbfbf66..e79497c5b619 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
@@ -68,22 +68,34 @@ static const u16 csc_coeff_rgb_out_eitu709[3][4] = {
 	{ 0x2000, 0x0000, 0x3b61, 0x7e25 }
 };
 
-static const u16 csc_coeff_rgb_in_eitu601[3][4] = {
+static const u16 csc_coeff_rgb_full_to_eitu601_full[3][4] = {
 	{ 0x2591, 0x1322, 0x074b, 0x0000 },
 	{ 0x6535, 0x2000, 0x7acc, 0x0200 },
 	{ 0x6acd, 0x7534, 0x2000, 0x0200 }
 };
 
-static const u16 csc_coeff_rgb_in_eitu709[3][4] = {
+static const u16 csc_coeff_rgb_full_to_eitu601_limited[3][4] = {
+	{ 0x2023, 0x105e, 0x063d, 0x0040 },
+	{ 0x688e, 0x1c00, 0x7b73, 0x0200 },
+	{ 0x6d73, 0x768e, 0x1c00, 0x0200 }
+};
+
+static const u16 csc_coeff_rgb_full_to_eitu709_full[3][4] = {
 	{ 0x2dc5, 0x0d9b, 0x049e, 0x0000 },
 	{ 0x62f0, 0x2000, 0x7d11, 0x0200 },
 	{ 0x6756, 0x78ab, 0x2000, 0x0200 }
 };
 
+static const u16 csc_coeff_rgb_full_to_eitu709_limited[3][4] = {
+	{ 0x2727, 0x0ba4, 0x03f3, 0x0040 },
+	{ 0x6692, 0x1c00, 0x7d6f, 0x0200 },
+	{ 0x6a6b, 0x7996, 0x1c00, 0x0200 }
+};
+
 static const u16 csc_coeff_rgb_full_to_rgb_limited[3][4] = {
-	{ 0x1b7c, 0x0000, 0x0000, 0x0020 },
-	{ 0x0000, 0x1b7c, 0x0000, 0x0020 },
-	{ 0x0000, 0x0000, 0x1b7c, 0x0020 }
+	{ 0x1b80, 0x0000, 0x0000, 0x0020 },
+	{ 0x0000, 0x1b80, 0x0000, 0x0020 },
+	{ 0x0000, 0x0000, 0x1b80, 0x0020 }
 };
 
 struct hdmi_vmode {
@@ -1105,7 +1117,7 @@ static bool is_csc_needed(struct dw_hdmi *hdmi)
 static void dw_hdmi_update_csc_coeffs(struct dw_hdmi *hdmi)
 {
 	const u16 (*csc_coeff)[3][4] = &csc_coeff_default;
-	bool is_input_rgb, is_output_rgb;
+	bool is_input_rgb, is_output_rgb, is_limited_range = true;
 	unsigned i;
 	u32 csc_scale = 1;
 
@@ -1119,9 +1131,13 @@ static void dw_hdmi_update_csc_coeffs(struct dw_hdmi *hdmi)
 			csc_coeff = &csc_coeff_rgb_out_eitu709;
 	} else if (is_input_rgb && !is_output_rgb) {
 		if (hdmi->hdmi_data.enc_out_encoding == V4L2_YCBCR_ENC_601)
-			csc_coeff = &csc_coeff_rgb_in_eitu601;
+			csc_coeff = is_limited_range ?
+				    &csc_coeff_rgb_full_to_eitu601_limited :
+				    &csc_coeff_rgb_full_to_eitu601_full;
 		else
-			csc_coeff = &csc_coeff_rgb_in_eitu709;
+			csc_coeff = is_limited_range ?
+				    &csc_coeff_rgb_full_to_eitu709_limited :
+				    &csc_coeff_rgb_full_to_eitu709_full;
 		csc_scale = 0;
 	} else if (is_input_rgb && is_output_rgb &&
 		   hdmi->hdmi_data.rgb_limited_range) {
-- 
2.43.0

